src/EditorFeatures/Core/Implementation/Formatting/FormatCommandHandler.cs

```diff
@@ -51,36 +54,38 @@ internal partial class FormatCommandHandler :
             _editorOperationsFactoryService = editorOperationsFactoryService;
         }
 
-        private void Format(ITextView textView, Document document, TextSpan? selectionOpt, CancellationToken cancellationToken)
+        private async Task Format(ITextView textView, Document document, TextSpan? selectionOpt, CancellationToken cancellationToken)
         {
             var formattingService = document.GetLanguageService<IEditorFormattingService>();
 
             using (Logger.LogBlock(FunctionId.CommandHandler_FormatCommand, KeyValueLogMessage.Create(LogType.UserAction, m => m["Span"] = selectionOpt?.Length ?? -1), cancellationToken))
             using (var transaction = new CaretPreservingEditTransaction(EditorFeaturesResources.Formatting, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
             {
-                var changes = formattingService.GetFormattingChangesAsync(document, selectionOpt, cancellationToken).WaitAndGetResult(cancellationToken);
-                if (changes.Count == 0)
+                var changes = await formattingService.GetFormattingChangesAsync(document, selectionOpt, cancellationToken).ConfigureAwait(false);
+                if (changes.Count > 0)
                 {
-                    return;
+                    ApplyChanges(document, changes, selectionOpt, cancellationToken);
+                    transaction.Complete();
```

Comment:
this does'nt seem safe at all.  You're on the UI thread, setting up an edit-transaction.  But you used ConfigureAwait(false).  

So this may be running on some random thread inappropriately.

Reply:
fixed